src/index.spec.ts

```
@@ -1,12 +1,182 @@
+      },
+      {
+        name: `['1','2','3'] when ['1','2','2','3','1'] is evaluated`,
+        actual: ['1', '2', '2', '3', '1'],
+        expected: ['1', '2', '3'],
+      },
+      {
+        name: 'the same array when the array only has one element',
+        actual: ['one'],
+        expected: ['one'],
+      },
+    ],
+  );
+
+  cases(
+    'should throw an error',
+    (opts) => {
+      expect(() => removeDuplicatesFromArray(opts.actual as any)).toThrowError(
+        opts.expected,
+      );
+    },
+    [
+      {
+        name: 'when there is no an array given',
+        actual: {},
+        expected: 'please provide an array of numbers or strings',
+      },
+    ],
+  );
+});
+
+describe('createRandomProduct', () => {
+  it('should return a random product when giving a creator role email', () => {
+    const creatorEmail = 'clark@kent.com';
+    const actual = createRandomProduct(creatorEmail);
+    expect(actual).not.toBeNull();
+  });
+
+  it('should throw an error when user does not have creator role', () => {
+    const noCreatorUser = 'diana@themyscira.com';
+    expect(() => createRandomProduct(noCreatorUser)).toThrowError(
+      'You are not allowed to create products',
+    );
+  });
+
+  it('should throw an error when passing an invalid email', () => {
+    const invalidEmail = 'invalidemail@fake.com';
+    expect(() =>
+      createRandomProduct(invalidEmail),
+    ).toThrowErrorMatchingInlineSnapshot(
+      `"Cannot read properties of undefined (reading 'role')"`,
+    );
+  });
+});
+
+describe('createProduct', () => {
+  const testProduct = {
+    name: 'motorcycle',
+    description: 'arch motorcycle',
+    price: 1000,
+    tags: ['at least one'],
+  };
+
+  it('should return a valid product', () => {
+    const actual = createProduct(testProduct);
+
+    expect(actual).toEqual({
+      id: expect.any(Number),
+      ...testProduct,
+    });
+  });
+
+  it('should throw an error whit a bad forming object', () => {
```

`should throw an error with a bad forming object`